## Preprocessing

In [31]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import pandas as pd 


In [32]:
#  Import and read the charity_data.csv.
application_df = pd.read_csv("charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [33]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=["EIN", "NAME"], axis=1)
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [34]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [35]:
# Look at APPLICATION_TYPE value counts for binning
#  YOUR CODE GOES HERE
application_df["APPLICATION_TYPE"].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T14        3
T25        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [36]:
# Choose a cutoff value and create a list of application types to be replaced
application_types_to_replace = ["T9", "T13", "T12", "T2", "T14", "T25", "T29", "T15", "T17"]

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [37]:
# Look at CLASSIFICATION value counts for binning
application_df["CLASSIFICATION"].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C1283        1
C2170        1
C1820        1
C1732        1
C1245        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [38]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#  YOUR CODE GOES HERE
application_df["CLASSIFICATION"].value_counts().head(45)

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C2300       32
C7200       32
C1240       30
C8000       20
C7120       18
C1500       16
C6000       15
C1800       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C1267        2
C3200        2
C1234        2
C1256        2
C1246        2
Name: CLASSIFICATION, dtype: int64

In [39]:
count_freq = dict(application_df["CLASSIFICATION"].value_counts())
classifications_to_replace = []

#drop under 1000

for element in count_freq:
    if count_freq[element] < 1000:
        classifications_to_replace.append(element)


In [40]:
# Choose a cutoff value and create a list of classifications to be replaced

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [51]:
# Convert categorical data to numeric with `pd.get_dummies`
df_dummies = pd.get_dummies(application_df,columns=["APPLICATION_TYPE","CLASSIFICATION","INCOME_AMT","AFFILIATION", "USE_CASE", "ORGANIZATION", "SPECIAL_CONSIDERATIONS"])
df_dummies

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0           1      5000              1                       0   
1           1    108590              1                       0   
2           1      5000              0                       0   
3           1      6692              1                       0   
4           1    142590              1                       0   
...       ...       ...            ...                     ...   
34294       1      5000              0                       0   
34295       1      5000              0                       0   
34296       1      5000              0                       0   
34297       1      5000              1                       0   
34298       1  36500179              0                       0   

       APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                         1                     0                    0   
1                         0                     0                    1   
2                         0                     0                    0   
3                         0                     0                    1   
4                         0                     0                    1   
...                     ...                   ...                  ...   
34294                     0                     0                    0   
34295                     0                     0                    0   
34296                     0                     0                    1   
34297                     0                     0                    0   
34298                     0                     0                    1   

       APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        0                    1                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    1                    0                    0  ...   
34295                    1                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    0                    1                    0  ...   
34298                    0                    0                    0  ...   

       USE_CASE_Heathcare  USE_CASE_Other  USE_CASE_Preservation  \
0                       0               0                      0   
1                       0               0                      1   
2                       0               0                      0   
3                       0               0                      1   
4                       1               0                      0   
...                   ...             ...                    ...   
34294                   0               0                      0   
34295                   0               0                      0   
34296                   0               0                      1   
34297                   0               0                      0   
34298                   0               0                      1   

       USE_CASE_ProductDev  ORGANIZATION_Association  \
0                        1                         1   
1                        0                         0   
2                        1                         1   
3                        0                         0   
4                        0                         0   
...                    ...                       ...   
34294                    1                         1   
34295                    1                         1   
34296                    0                         1   
34297                  

In [52]:
df = df_dummies

# Split our preprocessed data into our features and target arrays
Y_data = df["IS_SUCCESSFUL"]
X_data = df.drop("IS_SUCCESSFUL", axis=1)

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X_data, Y_data, random_state=50)

In [53]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [96]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_model = tf.keras.models.Sequential()

nn_model.add(tf.keras.layers.Dense(units=20, activation="relu", input_dim=43))

nn_model.add(tf.keras.layers.Dense(units=10, activation="relu"))
nn_model.add(tf.keras.layers.Dense(units=10, activation="relu"))
nn_model.add(tf.keras.layers.Dense(units=10, activation="relu"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_9 (Dense)             (None, 20)                880       
                                                                 
 dense_10 (Dense)            (None, 10)                210       
                                                                 
 dense_11 (Dense)            (None, 10)                110       
                                                                 
 dense_12 (Dense)            (None, 10)                110       
                                                                 
 dense_13 (Dense)            (None, 1)                 11        
                                                                 
Total params: 1,321
Trainable params: 1,321
Non-trainable params: 0
_________________________________________________________________


In [88]:
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh'])
    
    # Allow kerastuner to decide number of neurons in first layer
    #nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
    #    min_value=1,
    #    max_value=30,
    #    step=5), activation=activation, input_dim=43))

    nn_model.add(tf.keras.layers.Dense(units=5, activation="relu", input_dim=43))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 5)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=1,
            max_value=30,
            step=5),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [89]:
import kerastuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    max_epochs=20,
    hyperband_iterations=2)

INFO:tensorflow:Reloading Oracle from existing project .\untitled_project\oracle.json


In [90]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train,y_train,epochs=10,validation_data=(X_test,y_test))

Trial 61 Complete [00h 00m 23s]
val_accuracy: 0.5349271297454834

Best val_accuracy So Far: 0.5349271297454834
Total elapsed time: 00h 09m 20s
INFO:tensorflow:Oracle triggered exit


In [91]:
top_hyper = tuner.get_best_hyperparameters(3)
for param in top_hyper:
    print(param.values)

{'activation': 'tanh', 'first_units': 21, 'num_layers': 2, 'units_0': 11, 'tuner/epochs': 3, 'tuner/initial_epoch': 0, 'tuner/bracket': 2, 'tuner/round': 0, 'units_1': 1}
{'activation': 'relu', 'first_units': 26, 'num_layers': 4, 'units_0': 21, 'units_1': 11, 'tuner/epochs': 3, 'tuner/initial_epoch': 0, 'tuner/bracket': 2, 'tuner/round': 0, 'units_2': 1, 'units_3': 1}
{'activation': 'relu', 'first_units': 26, 'num_layers': 3, 'units_0': 16, 'units_1': 1, 'units_2': 16, 'units_3': 11, 'tuner/epochs': 3, 'tuner/initial_epoch': 0, 'tuner/bracket': 2, 'tuner/round': 0}


In [92]:
top_model = tuner.get_best_models(3)
for model in top_model:
    model_loss, model_accuracy = model.evaluate(X_test,y_test,verbose=2)
    print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.6907 - accuracy: 0.5349 - 259ms/epoch - 965us/step
Loss: 0.6907267570495605, Accuracy: 0.5349271297454834
268/268 - 0s - loss: 0.6908 - accuracy: 0.5349 - 348ms/epoch - 1ms/step
Loss: 0.6908339858055115, Accuracy: 0.5349271297454834
268/268 - 0s - loss: 0.7143 - accuracy: 0.5349 - 286ms/epoch - 1ms/step
Loss: 0.7142892479896545, Accuracy: 0.5349271297454834


In [97]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [98]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=20)

Epoch 1/20
804/804 [==============================] - 1s 709us/step - loss: 0.5872 - accuracy: 0.7104
Epoch 2/20
804/804 [==============================] - 1s 717us/step - loss: 0.5600 - accuracy: 0.7254
Epoch 3/20
804/804 [==============================] - 1s 775us/step - loss: 0.5562 - accuracy: 0.7272
Epoch 4/20
804/804 [==============================] - 1s 759us/step - loss: 0.5546 - accuracy: 0.7281
Epoch 5/20
804/804 [==============================] - 1s 708us/step - loss: 0.5531 - accuracy: 0.7293
Epoch 6/20
804/804 [==============================] - 1s 722us/step - loss: 0.5528 - accuracy: 0.7297
Epoch 7/20
804/804 [==============================] - 1s 723us/step - loss: 0.5518 - accuracy: 0.7303
Epoch 8/20
804/804 [==============================] - 1s 738us/step - loss: 0.5513 - accuracy: 0.7299
Epoch 9/20
804/804 [==============================] - 1s 785us/step - loss: 0.5511 - accuracy: 0.7305
Epoch 10/20
804/804 [==============================] - 1s 778us/step - loss: 0.550

In [78]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5452 - accuracy: 0.7383 - 203ms/epoch - 758us/step
Loss: 0.5452402830123901, Accuracy: 0.7383090257644653


In [79]:
# Export our model to HDF5 file
nn_model.save("AlphabetSoupCharity_Optimization.h5")

### REPORT


## Data Preprocessing

The target variable is "IS_SUCCESSFUL" a boolean variable indicating whether or not an investment by alphabet was successful
All of the other data in the dataframe are features, except meaningless text removed below
EIN and Company Name were removed from the data because they contained no information relevant to likelihood of loan success


## Compiling, Training, and Evaluating the Model

I used one input layer, three hidden layers, and one output layer. Output layer requires sigmoid to reduce to zero or one. Each hidden layer had 10 neurons to allow for some complexity.
Not quite able to achieve target model performance - I got to 73% vs. target of 75%
I tried using kerastuner to automatically increase performance but that didn't seem to help. I also added more hidden layers.

## Summary: Summarize the overall results of the deep learning model. Include a recommendation for how a different model could solve this classification problem, and explain your recommendation.
Overall this model performed reasonably well but certainly not perfect, I think with further iterations it could have gotten higher accuracy. I would also look at a logistic regression model for this approach, because we are essentially trying to predict the probability of an outcome (successful investment) being zero or one
